libraries/chain/db_block.cpp

```diff
@@ -301,9 +304,14 @@ processed_transaction database::push_proposal(const proposal_object& proposal)
       if( _undo_db.size() >= _undo_db.max_size() )
          _undo_db.set_max_size( _undo_db.size() + 1 );
       auto session = _undo_db.start_undo_session(true);
-      for( auto& op : proposal.proposed_transaction.operations )
+      for( auto& op : proposal_copy.proposed_transaction.operations )
          eval_state.operation_results.emplace_back(apply_operation(eval_state, op));
-      remove(proposal);
+      // The proposal object can be removed from database when processing,
+      // then be re-inserted to database when undo session destructs,
+      // in this case the address would be different, which means the reference would be invalidated already.
+      const proposal_object* po = find(proposal_id);
+      FC_ASSERT( po, "Proposal ${p} was removed unexpectedly", ("p",proposal_id) ); // make sure the proposal still exists
```

Comment:
If this assertion throws the session will be rolled back, which reconstructs the proposal.
This means the enclosing proposal_update succeeds, which leads to the proposal being removed again.
This will cause the enclosing push_proposal to throw at the same point, which again reconstructs it.

This alternating success/failure looks dirty to me. Does it mean the ultimate success or failure depends on the nesting depth?

Reply:
Actually this is generic defensive logic -- only tries to remove an object if it exists. In addition, to be defensive, locate the object again. The rationale is a proposal can be removed and then reconstructed not only due to self-approving (I believe we can make a test case for this). I do agree the self-approving proposal makes the situation more complicated, and we do need to avoid the situation in the first place as #1480 tries to do.

Reply:
> a proposal can be removed and then reconstructed not only due to self-approving (I believe we can make a test case for this)

Update:
I may be wrong. When pushing a proposal, the proposal itself should not be removed if we've avoided self-approving as well as loops.